src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs

```diff
@@ -44,56 +44,58 @@ public static class PackageDownloader
                 throw new ArgumentNullException(nameof(settings));
             }
 
-            using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token))
+            // While CancellationTokenSource has a dispose, if we call it before the tasks using
+            // the token are done we risk having these tasks fail with a object disposed exception so
+            // this should not be in a using unless we want to block until all tasks are done/cancled
+            var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
```

Comment:
You are right about the task encountering the ObjectDisposedException when the cancellation token source has been disposed. But, a linked token source should always be disposed. It is not safe to not dispose it. Refer https://msdn.microsoft.com/en-us/library/dd997364(v=vs.110).aspx, otherwise, it can cause a memory leak.

You are also right in that, it is by design, that we cannot wait for all the tasks.

There are 2 options as I see it for now. But, we have to think it through.

1. The task in PluginCredentialProvider can catch ObjectDisposedException and kill the process.
2. We don't use a LinkedTokenSource or use a LinkedTokenSource that is cached or something like that.


There are 2 options I can think of at this point